### DISCUSSION TITLE
Question about encapsulation

### DISCUSSION THREAD

**Discussion by paulorodrigues-mb** [fastify/fastify#3880] [General]
On the very first paragraph from Fastify encapsulation documentation says:

"The encapsulation context governs which [decorators](https://www.fastify.io/docs/latest/Reference/Decorators), registered [hooks](https://www.fastify.io/docs/latest/Reference/Hooks), and [plugins](https://www.fastify.io/docs/latest/Reference/Plugins) are available to [routes](https://www.fastify.io/docs/latest/Reference/Routes)."

Very nice, my question is:

Can I control which plugin are avaible for another plugin besides routes?

For exemple, I have to restrict my route to only access the business layer of my application and so the route can't directly access, for instance, to the database plugin. Only the business layer must have access to the database.

How can I accomplish that?

In my tests when then bussiness layer have access to the database plugin, the plugin is visible to the route too.

Hope to make my question clear, but feel free to ask any more questions.

Thanks

**Comment by jsumners** [comment]
Have you read the full document? Does it not answer this question?

**Reply by paulorodrigues-mb** [reply]
Yes I read, and actually no it doesn't, proprably my bad but if you could help to solve this, I really appreciate.

The code bellow has access to the database layer from route, but what I'm trying to accomplish is to make sure that only my bussiness layer has access to the database layer.

```javascript
const fastify = require('fastify')({ logger: true })
const fp = require('fastify-plugin')

async function Db (server, opts) {
  async function query (origin) {
    server.log.info('query executed ' + origin)
  }

  server.decorate('db', {
    query
  })
}

async function BusinessLayer (server, opts) {
  server.register(fp(Db))

  await server.after()

  const { db } = server

  async function create (origin) {
    server.log.info('business executed ' + origin)
    await db.query('from business')
  }

  server.decorate('business', {
    create
  })
}

fastify.register(fp(BusinessLayer))

fastify.route({
  path: '/',
  method: 'GET',
  async handler (req, reply) {
    await this.db.query('from route') // here, this code must not be possible, only business should have access to db layer
    await this.business.create('from route')

    return { ok: true }
  }
})

fastify
  .listen(3000)
  .catch(fastify.log.error)
```

How can I accomplish that?

Thanks

**Reply by mcollina** [reply]
You can do this:

```js

const fastify = require('fastify')({ logger: true })
const fp = require('fastify-plugin')

async function Db (server, opts) {
  async function query (origin) {
    server.log.info('query executed ' + origin)
  }

  server.decorate('db', {
    query
  })
}

async function BusinessLayer (server, opts) {
  server.register(fp(Db))

  await server.after()

  const { db } = server

  async function create (origin) {
    server.log.info('business executed ' + origin)
    await db.query('from business')
  }

  server.decorate('business', {
    create
  })

fastify.route({
  path: '/',
  method: 'GET',
  async handler (req, reply) {
    await this.db.query('from route') // here, this code must not be possible, only business should have access to db layer
    await this.business.create('from route')

    return { ok: true }
  }
})

}

fastify.register(BusinessLayer)

fastify
  .listen(3000)
  .catch(fastify.log.error)

```

**Comment by Eomm** [comment]
Starting from your code:

```diff
- fastify.register(fp(BusinessLayer))
+ fastify.register(BusinessLayer)
```

the docs say that `fastify-plugin` breaks the encapsulation.

The other concepts you get are correct

**Reply by paulorodrigues-mb** [reply]
Hi Manuel, thank you for your help.

If I don't break encapsulation there, I can't have access to business layer from the route.

I must have access to business layer from route, but not to the database layer.

**Reply by Eomm** [reply]
Use `server.route({` instead of `fastify.route({`

**Reply by Eomm** [reply]
This plugin will help you to understand the encapsulation visibility
https://github.com/Eomm/fastify-overview
